{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3d9aec7c",
   "metadata": {},
   "source": [
    "# Geometrical Pairwise Analysis of Biological Sequence Following Kabirian-based Optinalysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6c471c21",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''                            \n",
    "                                # USER GUIDE\n",
    "                            #*******************#\n",
    "# Introduction: \n",
    "    Geometrical pairwise analysis (GPA) of biological sequences following Kabirian-based optinalysis looks at two biological \n",
    "    sequences (i.e., DNA, RNA, and amino acid sequences) as isoreflective as a mirror-like reflection of each other \n",
    "    about a central point. It estimates the percentage similarity between two sequences. \n",
    "\n",
    "# Input guide: pairwise_analysis([sequence_x, sequence_y, encoding_scheme, pairing, print_result]) \n",
    "    # Input options:  \n",
    "        # for sequences: aligned strings of biological sequences (i.e., DNA, RNA, or protein). \n",
    "        # for encoding_scheme: \"seq_type:DNA\", \"seq_type:RNA\", or \"seq_type:Protein\".\n",
    "        # for pairing: \"pairing:H_H\", or \"pairing:T_T\". \n",
    "        # for print_result: \"print:kc\", \"print:psim\", \"print:pdsim\", \"print:kcalt1\", \"print:kcalt2\", \"print:kcalt\", or \"all\".\n",
    "    # Note:\n",
    "        encoding_scheme input is what convert each biological sequence into a numerical representation using \n",
    "        a specific encoding scheme. This encoding captures the features of the sequences such as base or \n",
    "        amino acid compositions, and aligned gabs of the sequences. \n",
    "    \n",
    "# Example 1: geometrical pairwise_analysis of DNA sequences \n",
    "    # print(\"Kabirian coefficient =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:kc\"])) \n",
    "    # print(\"Probability of similarity =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:psim\"])) \n",
    "    # print(\"Probability of dissimilarity =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:pdsim\"])) \n",
    "    # print(\"Alt1. Kabirian coefficient =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:kcalt1\"])) \n",
    "    # print(\"Alt2. Kabirian coefficient =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:kcalt2\"])) \n",
    "    # print(\"Alt. Kabirian coefficient =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:kcalt\"])) \n",
    "    # print(\"All the estimates =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:all\"])) \n",
    "\n",
    "# Example 2: geometrical pairwise_analysis of RNA sequences \n",
    "    # print(\"Kabirian coefficient =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:kc\"])) \n",
    "    # print(\"Probability of similarity =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:psim\"])) \n",
    "    # print(\"Probability of dissimilarity =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:pdsim\"])) \n",
    "    # print(\"Alt1. Kabirian coefficient =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:kcalt1\"])) \n",
    "    # print(\"Alt2. Kabirian coefficient =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:kcalt2\"])) \n",
    "    # print(\"Alt. Kabirian coefficient =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:kcalt\"])) \n",
    "    # print(\"All the estimates =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:all\"])) \n",
    "\n",
    "# Example 3: geometrical pairwise_analysis of protein sequences \n",
    "    # print(\"Kabirian coefficient =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:kc\"])) \n",
    "    # print(\"Probability of similarity =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:psim\"])) \n",
    "    # print(\"Probability of dissimilarity =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:pdsim\"])) \n",
    "    # print(\"Alt1. Kabirian coefficient =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:kcalt1\"])) \n",
    "    # print(\"Alt2. Kabirian coefficient =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:kcalt2\"])) \n",
    "    # print(\"Alt. Kabirian coefficient =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:kcalt\"])) \n",
    "    # print(\"All the estimates =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:all\"])) \n",
    "\n",
    "#******************************************************************************************************#\n",
    "'''\n",
    "# Importing necessary libraries\n",
    "import numpy\n",
    "import numpy as np\n",
    "\n",
    "# Function for performing isomorphic optinalysis\n",
    "def pairwise_analysis(instruction_list):\n",
    "    # Extracting data from the instruction_list\n",
    "    sequence_x = instruction_list[0]  # Data from the first list\n",
    "    sequence_y = instruction_list[1]  # Data from the second list\n",
    "    encoding_scheme = instruction_list[2] # Encoding scheme\n",
    "    pairing = instruction_list[3] # Type of pairing\n",
    "    print_result = instruction_list[4] # Type of result(s) to print\n",
    "    \n",
    "    # Numerical encoding scheme for DNA sequences\n",
    "    def DNA_encoding(seq):\n",
    "        encoding = {'c': 111, 't': 126, 'a': 135, 'g': 151, '-': 0, 'C': 111, 'T': 126, 'A': 135, 'G': 151,}\n",
    "        return [encoding.get(n, 0) for n in seq]\n",
    "\n",
    "    # Numerical encoding scheme for RNA sequences\n",
    "    def RNA_encoding(seq):\n",
    "        encoding = {'c': 111, 'u': 112, 'a': 135, 'g': 151, '-': 0, 'C': 111, 'U': 112, 'A': 135, 'G': 151}\n",
    "        return [encoding.get(n, 0) for n in seq]\n",
    "\n",
    "    # Numerical encoding scheme for protein sequences\n",
    "    def protein_encoding(seq):\n",
    "        encoding = {'G': 75, 'A': 89, 'S': 105, 'P': 115, 'V': 117, 'T': 119, 'C': 121, 'I': 131, 'L': 131, 'N': 132, \n",
    "                    'D': 133, 'Q': 146, 'K': 146, 'E': 147, 'M': 149, 'H': 155, 'F': 165, 'R': 174, 'Y': 181, 'W': 204,\n",
    "                   'g': 75, 'a': 89, 's': 105, 'p': 115, 'v': 117, 't': 119, 'c': 121, 'i': 131, 'l': 131, 'n': 132, \n",
    "                    'd': 133, 'q': 146, 'k': 146, 'e': 147, 'm': 149, 'h': 155, 'f': 165, 'r': 174, 'y': 181, 'W': 204, '-': 0}\n",
    "        return [encoding.get(n, 0) for n in seq]\n",
    "\n",
    "    # Function for computing Kabirian coefficient (kc) of isomorphic optinalysis\n",
    "    def kc_isomorphic_optinalysis(instruction_list):\n",
    "        # Extracting data from the instruction_list\n",
    "        data_x = instruction_list[0]  # Data from the first list\n",
    "        data_y = instruction_list[1]  # Data from the second list\n",
    "        pairing = instruction_list[2]  # Type of pairing\n",
    "\n",
    "        # Generating a list of optiscale values from 0.01 to 2 times the length of data_x\n",
    "        optiscale = [p / 100 for p in range(1, (2 * len(data_x) + 2))]\n",
    "\n",
    "        # Calculating the mid-point of the optiscale values\n",
    "        mid_optiscale = (optiscale[0] * len(data_x)) + optiscale[0]\n",
    "\n",
    "        # Generating the isoreflective list based on the pairing type\n",
    "        if pairing == \"pairing:H_H\":\n",
    "            isoreflective_list = data_x + [0] + (data_y[::-1])  \n",
    "        elif pairing == \"pairing:T_T\":\n",
    "            isoreflective_list = (data_x[::-1]) + [0] + data_y\n",
    "        else:\n",
    "            print('Invalid command. Please, use \"pairing:H_H\", or \"pairing:T_T\" to command Head-to-head, or Tail-to-tail pairing respectivelly')\n",
    "\n",
    "        # Calculating the dot product of isoreflective_list and optiscale\n",
    "        sum_of_scalements = np.dot(isoreflective_list, optiscale)\n",
    "\n",
    "        # Calculating the kc_optinalysis using the calculated values\n",
    "        kc_optinalysis = (mid_optiscale * sum(isoreflective_list)) / sum_of_scalements\n",
    "\n",
    "        # Returning the calculated kc_optinalysis value\n",
    "        return kc_optinalysis\n",
    "\n",
    "    # Function for translating Kabirian coefficient (kc) to percentage similarity (psim)\n",
    "    def psim(kc, num_of_dimensions):\n",
    "        # Check if kc is within the valid range [0, 1]\n",
    "        if 0 <= kc <= 1:\n",
    "            # Calculate psim using the formula for kc in [0, 1]\n",
    "            psim = ((num_of_dimensions + 1) - kc * ((2 * num_of_dimensions) + 1)) / (kc - (num_of_dimensions + 1))\n",
    "        else:\n",
    "            # Calculate psim using the formula for kc outside the range [0, 1]\n",
    "            psim = ((num_of_dimensions + 1) - kc) / (kc * ((2 * num_of_dimensions) + 1) - (num_of_dimensions + 1))\n",
    "        return psim  # Return the calculated psim value\n",
    "    \n",
    "    # Function for translating percentage similarity (psim) to percentage dissimilarity (pdsim)\n",
    "    def pdsim(psim):\n",
    "        # Check if psim is within the valid range [0, 1]\n",
    "        if 0 <= psim <= 1:\n",
    "            # Calculate pdsim when psim is in [0, 1]\n",
    "            pdsim = 1 - psim\n",
    "        else:\n",
    "            # Calculate pdsim when psim is outside [0, 1]\n",
    "            pdsim = -1 - psim\n",
    "        return pdsim\n",
    "\n",
    "    # Function for translating percentage similarity (psim) to alternative Kabirian coefficient (kcalt)\n",
    "    def kc_alt(kc, psim, num_of_dimensions):\n",
    "        if 0 <= kc <= 1:\n",
    "            # Calculate kc_alt when kc is in [0, 1]\n",
    "            kc_alt = ((num_of_dimensions + 1) * (psim + 1)) / (((2 * num_of_dimensions) + 1) * psim + 1)\n",
    "        else:\n",
    "            # Calculate kc_alt when kc is outside [0, 1]\n",
    "            kc_alt = ((num_of_dimensions + 1) * (psim + 1)) / (psim + ((2 * num_of_dimensions) + 1))\n",
    "        return kc_alt\n",
    "    \n",
    "    # Function for translating percentage similarity (psim) to alternative Kabirian coefficient (kcalt1)\n",
    "    def kc_alt1(psim, num_of_dimensions):\n",
    "        # Calculate kc_alt1\n",
    "        kc_alt1 = ((num_of_dimensions + 1) * (psim + 1)) / (psim + ((2 * num_of_dimensions) + 1))\n",
    "        return kc_alt1\n",
    "    \n",
    "    # Function for translating percentage similarity (psim) to alternative Kabirian coefficient (kcalt2)\n",
    "    def kc_alt2(psim, num_of_dimensions):\n",
    "        # Calculate kc_alt2\n",
    "        kc_alt2 = ((num_of_dimensions + 1) * (psim + 1)) / (((2 * num_of_dimensions) + 1) * psim + 1)\n",
    "        return kc_alt2\n",
    "\n",
    "    # Encoding Sequences Based on the Selected Encoding Scheme\n",
    "    if encoding_scheme == \"seq_type:DNA\":\n",
    "        seq_1 = DNA_encoding(sequence_x)\n",
    "        seq_2 = DNA_encoding(sequence_y)\n",
    "    elif encoding_scheme == \"seq_type:RNA\":\n",
    "        seq_1 = RNA_encoding(sequence_x)\n",
    "        seq_2 = RNA_encoding(sequence_y)\n",
    "    elif encoding_scheme == \"seq_type:protein\":\n",
    "        seq_1 = protein_encoding(sequence_x)\n",
    "        seq_2 = protein_encoding(sequence_y)\n",
    "    else:\n",
    "        print('Invalid command. Please use \"seq_type:DNA\", \"seq_type:RNA\", or \"seq_type:protein\" to command encoding_scheme.')\n",
    "\n",
    "    # Perform Isomorphic Optinalysis\n",
    "    kc_result = kc_isomorphic_optinalysis([seq_1, seq_2, pairing])\n",
    "    num_of_dimensions = len(sequence_x)\n",
    "\n",
    "    # Calculate All Other Estimates\n",
    "    psim_result = psim(kc_result, num_of_dimensions)\n",
    "    pdsim_result = pdsim(psim_result)\n",
    "    kc_alt1_result = kc_alt1(psim_result, num_of_dimensions)\n",
    "    kc_alt2_result = kc_alt2(psim_result, num_of_dimensions)\n",
    "    kc_alt_result = kc_alt(kc_result, psim_result, num_of_dimensions)\n",
    "\n",
    "    # Create a dictionary containing all the estimates\n",
    "    all_estimates = {\n",
    "        \"kc\": kc_result,\n",
    "        \"psim\": psim_result,\n",
    "        \"pdsim\": pdsim_result,\n",
    "        \"kc_alt1\": kc_alt1_result,\n",
    "        \"kc_alt2\": kc_alt2_result\n",
    "    }\n",
    "\n",
    "    # Select the result to be printed based on the print_result input\n",
    "    if print_result == \"print:kc\":\n",
    "        result = kc_result\n",
    "    elif print_result == \"print:psim\":\n",
    "        result = psim_result\n",
    "    elif print_result == \"print:pdsim\":\n",
    "        result = pdsim_result\n",
    "    elif print_result == \"print:kcalt1\":\n",
    "        result = kc_alt1_result\n",
    "    elif print_result == \"print:kcalt2\":\n",
    "        result = kc_alt2_result\n",
    "    elif print_result == \"print:kcalt\":\n",
    "        result = kc_alt_result\n",
    "    elif print_result == \"print:all\":\n",
    "        result = all_estimates\n",
    "    else:\n",
    "        result = 'Invalid command. Please use \"print:kc\", \"print:psim\", \"print:pdsim\", \"print:kcalt1\", \"print:kcalt2\", \"print:kcalt\", or \"print:all\" to command print_result'\n",
    "    \n",
    "    return result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61257269",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f58adbc0",
   "metadata": {},
   "source": [
    "# Practical examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "536ada60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Example 1: geometrical pairwise_analysis of DNA sequences\n",
      "Kabirian coefficient = 1.0035643017098081\n",
      "Probability of similarity = 0.9927651307349952\n",
      "Probability of dissimilarity = 0.0072348692650048285\n",
      "Alt1. Kabirian coefficient = 0.9964609269385202\n",
      "Alt2. Kabirian coefficient = 1.003564301709808\n",
      "Alt. Kabirian coefficient = 0.9964609269385202\n",
      "All the estimates = {'kc': 1.0035643017098081, 'psim': 0.9927651307349952, 'pdsim': 0.0072348692650048285, 'kc_alt1': 0.9964609269385202, 'kc_alt2': 1.003564301709808}\n",
      "******************************************************************\n",
      "Example 2: geometrical pairwise_analysis of RNA sequences\n",
      "Kabirian coefficient = 0.9962183715542534\n",
      "Probability of similarity = 0.9922693203899832\n",
      "Probability of dissimilarity = 0.007730679610016833\n",
      "Alt1. Kabirian coefficient = 0.9962183715542537\n",
      "Alt2. Kabirian coefficient = 1.0038104478416445\n",
      "Alt. Kabirian coefficient = 1.0038104478416445\n",
      "All the estimates = {'kc': 0.9962183715542534, 'psim': 0.9922693203899832, 'pdsim': 0.007730679610016833, 'kc_alt1': 0.9962183715542537, 'kc_alt2': 1.0038104478416445}\n",
      "******************************************************************\n",
      "Example 3: geometrical pairwise_analysis of protein sequences\n",
      "Kabirian coefficient = 0.9709266753984838\n",
      "Probability of similarity = 0.9407248843064939\n",
      "Probability of dissimilarity = 0.059275115693506075\n",
      "Alt1. Kabirian coefficient = 0.9709266753984839\n",
      "Alt2. Kabirian coefficient = 1.0308682074305042\n",
      "Alt. Kabirian coefficient = 1.0308682074305042\n",
      "All the estimates = {'kc': 0.9709266753984838, 'psim': 0.9407248843064939, 'pdsim': 0.059275115693506075, 'kc_alt1': 0.9709266753984839, 'kc_alt2': 1.0308682074305042}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# For DNA sequences that has been aligned \n",
    "sequence_1 = 'AGTAAGCAAACCGAACAGGTGTTCCTAGCATAATACACG------'\n",
    "sequence_2 = 'AGTAAGCAAAAGGTGCAGGTGTTCC---GAAAATACACGTAGTGC'\n",
    "\n",
    "# For RNA sequences that has been from a converted aligned DNA sequences \n",
    "sequence_3 = 'AGUAAGCAAACCGAACAGGUGUUCCUAGCAUAAUACACG------'\n",
    "sequence_4 = 'AGUAAGCAAAAGGUGCAGGUGUUCC---GAAAAUACACGUAGUGC'\n",
    "\n",
    "# For protein sequences that has been aligned \n",
    "sequence_5 = 'GDQ-----SHFRYWKEMNDQSHFRHFRYKEMWILNDQK-----MEDNDQE'\n",
    "sequence_6 = 'GDQFRYNYMHFRY----NDQKEM-YWKEKEMWILNDQKDNDQEMEDNDQE'\n",
    "\n",
    "print(\"Example 1: geometrical pairwise_analysis of DNA sequences\") \n",
    "print(\"Kabirian coefficient =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:kc\"])) \n",
    "print(\"Probability of similarity =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:psim\"])) \n",
    "print(\"Probability of dissimilarity =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:pdsim\"])) \n",
    "print(\"Alt1. Kabirian coefficient =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:kcalt1\"])) \n",
    "print(\"Alt2. Kabirian coefficient =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:kcalt2\"])) \n",
    "print(\"Alt. Kabirian coefficient =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:kcalt\"])) \n",
    "print(\"All the estimates =\", pairwise_analysis([sequence_1, sequence_2, \"seq_type:DNA\", \"pairing:H_H\", \"print:all\"])) \n",
    "\n",
    "print(\"******************************************************************\")\n",
    "\n",
    "print(\"Example 2: geometrical pairwise_analysis of RNA sequences\") \n",
    "print(\"Kabirian coefficient =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:kc\"])) \n",
    "print(\"Probability of similarity =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:psim\"])) \n",
    "print(\"Probability of dissimilarity =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:pdsim\"])) \n",
    "print(\"Alt1. Kabirian coefficient =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:kcalt1\"])) \n",
    "print(\"Alt2. Kabirian coefficient =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:kcalt2\"])) \n",
    "print(\"Alt. Kabirian coefficient =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:kcalt\"])) \n",
    "print(\"All the estimates =\", pairwise_analysis([sequence_4, sequence_3, \"seq_type:RNA\", \"pairing:H_H\", \"print:all\"])) \n",
    "\n",
    "print(\"******************************************************************\")\n",
    "\n",
    "print(\"Example 3: geometrical pairwise_analysis of protein sequences\") \n",
    "print(\"Kabirian coefficient =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:kc\"])) \n",
    "print(\"Probability of similarity =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:psim\"])) \n",
    "print(\"Probability of dissimilarity =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:pdsim\"])) \n",
    "print(\"Alt1. Kabirian coefficient =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:kcalt1\"])) \n",
    "print(\"Alt2. Kabirian coefficient =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:kcalt2\"])) \n",
    "print(\"Alt. Kabirian coefficient =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:kcalt\"])) \n",
    "print(\"All the estimates =\", pairwise_analysis([sequence_5, sequence_6, \"seq_type:protein\", \"pairing:H_H\", \"print:all\"])) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68047968",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4e75896e",
   "metadata": {},
   "source": [
    "# Well-structured Datasets (Sequences) Used in the Article Publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4021a148",
   "metadata": {},
   "outputs": [],
   "source": [
    "# well-structured DNA sequences\n",
    "\n",
    "S0 = 'CTAGCTAGCTAG'\n",
    "S1 = 'GTAGCTAGCTAG'\n",
    "S2 = 'CAAGCTAGCTAG'\n",
    "S3 = 'CTTGCTAGCTAG'\n",
    "S4 = 'CTACCTAGCTAG'\n",
    "S5 = 'CTAGGTAGCTAG'\n",
    "S6 = 'CTAGCAAGCTAG'\n",
    "S7 = 'CTAGCTTGCTAG'\n",
    "S8 = 'CTAGCTACCTAG'\n",
    "S9 = 'CTAGCTAGGTAG'\n",
    "S10 = 'CTAGCTAGCAAG'\n",
    "S11 = 'CTAGCTAGCTTG'\n",
    "S12 = 'CTAGCTAGCTAC'\n",
    "S13 = 'GTAGCTAGCTAG'\n",
    "S14 = 'ATAGCTAGCTAG'\n",
    "S15 = 'TTAGCTAGCTAG'\n",
    "S16 = 'CTAGCGAGCTAG'\n",
    "S17 = 'CTAGCAAGCTAG'\n",
    "S18 = 'CTAGCCAGCTAG'\n",
    "S19 = 'CTAGCTAGCTGG'\n",
    "S20 = 'CTAGCTAGCTTG'\n",
    "S21 = 'CTAGCTAGCTCG'\n",
    "S22 = '-TAGCTAGCTAG'\n",
    "S23 = 'C-AGCTAGCTAG'\n",
    "S24 = 'CT-GCTAGCTAG'\n",
    "S25 = 'CTA-CTAGCTAG'\n",
    "S26 = 'CTAG-TAGCTAG'\n",
    "S27 = 'CTAGC-AGCTAG'\n",
    "S28 = 'CTAGCT-GCTAG'\n",
    "S29 = 'CTAGCTA-CTAG'\n",
    "S30 = 'CTAGCTAG-TAG'\n",
    "S31 = 'CTAGCTAGC-AG'\n",
    "S32 = 'CTAGCTAGCT-G'\n",
    "S33 = 'CTAGCTAGCTA-'\n",
    "S34 = 'CTAGCTAGCT--'\n",
    "S35 = 'CTAGCTAGC---'\n",
    "S36 = 'CTAGCTAG----'\n",
    "S37 = 'CTAGCTA-----'\n",
    "S38 = 'CTAGCT------'\n",
    "S39 = 'CTAGC-------'\n",
    "S40 = 'CTAG--------'\n",
    "S41 = '--------CTAG'\n",
    "S42 = '-------GCTAG'\n",
    "S43 = '------AGCTAG'\n",
    "S44 = '-----TAGCTAG'\n",
    "S45 = '----CTAGCTAG'\n",
    "S46 = '---GCTAGCTAG'\n",
    "S47 = '--AGCTAGCTAG'\n",
    "S48 = 'GTA-CTGCCT-G'\n",
    "S49 = 'C-GG-TAGATAT'\n",
    "S50 = '-TAGGTAACCA-'\n",
    "S51 = 'C--GCTAAGTAA'\n",
    "S52 = 'TCAGCGAGCT--'\n",
    "S53 = 'CCAGCGC--TAG'\n",
    "S54 = 'CCAGCCG--TAG'\n",
    "S55 = 'CAAGCCG--TAG'\n",
    "S56 = 'CGAGCCG--TAG'\n",
    "S57 = 'C--ACTAGGGAG'\n",
    "S58 = 'C--ACTAGCGAT'\n",
    "S59 = 'C--TCTAGCATG'\n",
    "S60 = 'C--CCTAGCAGT'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "919b4a5b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b73ad9cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   SEQ ID           SEQ        KC      PSIM     PDSIM    KCALT1    KCALT2\n",
      "0      S1  GTAGCTAGCTAG  0.988515  0.975140  0.024860  0.988515  1.011755\n",
      "1      S2  CAAGCTAGCTAG  0.997586  0.994771  0.005229  0.997586  1.002426\n",
      "2      S3  CTTGCTAGCTAG  1.002217  0.995218  0.004782  0.997792  1.002217\n",
      "3      S4  CTACCTAGCTAG  1.009019  0.980818  0.019182  0.991140  1.009019\n",
      "4      S5  CTAGGTAGCTAG  0.992314  0.983358  0.016642  0.992314  1.007806\n",
      "5      S6  CTAGCAAGCTAG  0.998462  0.996669  0.003331  0.998462  1.001542\n",
      "6      S7  CTAGCTTGCTAG  1.001329  0.997128  0.002872  0.998674  1.001329\n",
      "7      S8  CTAGCTACCTAG  1.004991  0.989298  0.010702  0.995059  1.004991\n",
      "8      S9  CTAGCTAGGTAG  0.996142  0.991644  0.008356  0.996142  1.003888\n",
      "9     S10  CTAGCTAGCAAG  0.999340  0.998571  0.001429  0.999340  1.000660\n",
      "10    S11  CTAGCTAGCTTG  1.000443  0.999042  0.000958  0.999558  1.000443\n",
      "11    S12  CTAGCTAGCTAC  1.000994  0.997850  0.002150  0.999008  1.000994\n",
      "12    S13  GTAGCTAGCTAG  0.988515  0.975140  0.024860  0.988515  1.011755\n",
      "13    S14  ATAGCTAGCTAG  0.993042  0.984934  0.015066  0.993042  1.007056\n",
      "14    S15  TTAGCTAGCTAG  0.995628  0.990530  0.009470  0.995628  1.004411\n",
      "15    S16  CTAGCGAGCTAG  0.995762  0.990821  0.009179  0.995762  1.004274\n",
      "16    S17  CTAGCAAGCTAG  0.998462  0.996669  0.003331  0.998462  1.001542\n",
      "17    S18  CTAGCCAGCTAG  1.002593  0.994412  0.005588  0.997420  1.002593\n",
      "18    S19  CTAGCTAGCTGG  0.999220  0.998310  0.001690  0.999220  1.000781\n",
      "19    S20  CTAGCTAGCTTG  1.000443  0.999042  0.000958  0.999558  1.000443\n",
      "20    S21  CTAGCTAGCTCG  1.001187  0.997434  0.002566  0.998816  1.001187\n",
      "21    S22  -TAGCTAGCTAG  1.035035  0.929254  0.070746  0.967259  1.035035\n",
      "22    S23  C-AGCTAGCTAG  1.036696  0.926139  0.073861  0.965813  1.036696\n",
      "23    S24  CT-GCTAGCTAG  1.035819  0.927780  0.072220  0.966575  1.035819\n",
      "24    S25  CTA-CTAGCTAG  1.036267  0.926941  0.073059  0.966186  1.036267\n",
      "25    S26  CTAG-TAGCTAG  1.023087  0.952273  0.047727  0.977932  1.023087\n",
      "26    S27  CTAGC-AGCTAG  1.023044  0.952358  0.047642  0.977971  1.023044\n",
      "27    S28  CTAGCT-GCTAG  1.021188  0.956033  0.043967  0.979673  1.021188\n",
      "28    S29  CTAGCTA-CTAG  1.019829  0.958742  0.041258  0.980928  1.019829\n",
      "29    S30  CTAGCTAG-TAG  1.011412  0.975849  0.024151  0.988843  1.011412\n",
      "30    S31  CTAGCTAGC-AG  1.009748  0.979300  0.020700  0.990439  1.009748\n",
      "31    S32  CTAGCTAGCT-G  1.006964  0.985126  0.014874  0.993131  1.006964\n",
      "32    S33  CTAGCTAGCTA-  1.003904  0.991610  0.008390  0.996126  1.003904\n",
      "33    S34  CTAGCTAGCT--  1.011485  0.975696  0.024304  0.988772  1.011485\n",
      "34    S35  CTAGCTAGC---  1.023066  0.952314  0.047686  0.977951  1.023066\n",
      "35    S36  CTAGCTAG----  1.037952  0.923796  0.076204  0.964726  1.037952\n",
      "36    S37  CTAGCTA-----  1.066525  0.873408  0.126592  0.941287  1.066525\n",
      "37    S38  CTAGCT------  1.102224  0.817401  0.182599  0.915128  1.102224\n",
      "38    S39  CTAGC-------  1.147902  0.755029  0.244971  0.885861  1.147902\n",
      "39    S40  CTAG--------  1.202405  0.691530  0.308470  0.855920  1.202405\n",
      "40    S41  --------CTAG  1.475317  0.482549  0.517451  0.756327  1.475317\n",
      "41    S42  -------GCTAG  1.378546  0.541448  0.458552  0.784561  1.378546\n",
      "42    S43  ------AGCTAG  1.303453  0.597179  0.402821  0.811157  1.303453\n",
      "43    S44  -----TAGCTAG  1.240691  0.652669  0.347331  0.837523  1.240691\n",
      "44    S45  ----CTAGCTAG  1.189968  0.705110  0.294890  0.862336  1.189968\n",
      "45    S46  ---GCTAGCTAG  1.127563  0.781643  0.218357  0.898366  1.127563\n",
      "46    S47  --AGCTAGCTAG  1.077668  0.855156  0.144844  0.932774  1.077668\n",
      "47    S48  GTA-CTGCCT-G  1.034776  0.929743  0.070257  0.967486  1.034776\n",
      "48    S49  C-GG-TAGATAT  1.056964  0.889672  0.110328  0.948862  1.056964\n",
      "49    S50  -TAGGTAACCA-  1.035563  0.928261  0.071739  0.966798  1.035563\n",
      "50    S51  C--GCTAAGTAA  1.076709  0.856698  0.143302  0.933494  1.076709\n",
      "51    S52  TCAGCGAGCT--  1.006215  0.986707  0.013293  0.993862  1.006215\n",
      "52    S53  CCAGCGC--TAG  1.037159  0.925274  0.074726  0.965412  1.037159\n",
      "53    S54  CCAGCCG--TAG  1.038317  0.923118  0.076882  0.964411  1.038317\n",
      "54    S55  CAAGCCG--TAG  1.030460  0.937942  0.062058  0.971290  1.030460\n",
      "55    S56  CGAGCCG--TAG  1.025358  0.947814  0.052186  0.975866  1.025358\n",
      "56    S57  C--ACTAGGGAG  1.074732  0.859894  0.140106  0.934985  1.074732\n",
      "57    S58  C--ACTAGCGAT  1.082356  0.847694  0.152306  0.929291  1.082356\n",
      "58    S59  C--TCTAGCATG  1.086508  0.841188  0.158812  0.926252  1.086508\n",
      "59    S60  C--CCTAGCAGT  1.090541  0.834959  0.165041  0.923341  1.090541\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Analysis (optinalysis) of the well-structured DNA sequences\n",
    "lab = \"S1, S2, S3, S4, S5, S6, S7, S8, S9, S10, S11, S12, S13, S14, S15, S16, S17, S18, S19, S20, S21, S22, S23, S24, S25, S26, S27, S28, S29, S30, S31, S32, S33, S34, S35, S36, S37, S38, S39, S40, S41, S42, S43, S44, S45, S46, S47, S48, S49, S50, S51, S52, S53, S54, S55, S56, S57, S58, S59, S60\"\n",
    "DNA_sequences = [S1, S2, S3, S4, S5, S6, S7, S8, S9, S10, S11, S12, S13, S14, S15, S16, S17, S18, S19, S20, S21, S22, S23, S24, S25, S26, S27, S28, S29, S30, S31, S32, S33, S34, S35, S36, S37, S38, S39, S40, S41, S42, S43, S44, S45, S46, S47, S48, S49, S50, S51, S52, S53, S54, S55, S56, S57, S58, S59, S60]\n",
    "kc_results = [pairwise_analysis([S0, seq, \"seq_type:DNA\", \"pairing:H_H\", \"print:kc\"]) for seq in DNA_sequences]\n",
    "psim_results = [pairwise_analysis([S0, seq, \"seq_type:DNA\", \"pairing:H_H\", \"print:psim\"]) for seq in DNA_sequences]\n",
    "pdsim_results = [pairwise_analysis([S0, seq, \"seq_type:DNA\", \"pairing:H_H\", \"print:pdsim\"]) for seq in DNA_sequences]\n",
    "kcalt1_results = [pairwise_analysis([S0, seq, \"seq_type:DNA\", \"pairing:H_H\", \"print:kcalt1\"]) for seq in DNA_sequences]\n",
    "kcalt2_results = [pairwise_analysis([S0, seq, \"seq_type:DNA\", \"pairing:H_H\", \"print:kcalt2\"]) for seq in DNA_sequences]\n",
    "\n",
    "seq_ID = ['S{}'.format(i) for i in range(1, len(DNA_sequences) + 1)]\n",
    "DNA_seq = [str(i) for i in DNA_sequences]\n",
    "df = pd.DataFrame({\"SEQ ID\":seq_ID, \"SEQ\":DNA_seq, \"KC\":kc_results, \"PSIM\":psim_results, \"PDSIM\":pdsim_results, \"KCALT1\":kcalt1_results, \"KCALT2\":kcalt2_results})\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fb069db",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d762cb67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# well-structured protein sequences\n",
    "\n",
    "S0 = 'GASPCLDQMFRY'\n",
    "S1 = 'WASPCLDQMFRY'\n",
    "S2 = 'GWSPCLDQMFRY'\n",
    "S3 = 'GAWPCLDQMFRY'\n",
    "S4 = 'GASWCLDQMFRY'\n",
    "S5 = 'GASPWLDQMFRY'\n",
    "S6 = 'GASPCWDQMFRY'\n",
    "S7 = 'GASPCLWQMFRY'\n",
    "S8 = 'GASPCLDWMFRY'\n",
    "S9 = 'GASPCLDQWFRY'\n",
    "S10 = 'GASPCLDQMWRY'\n",
    "S11 = 'GASPCLDQMFWY'\n",
    "S12 = 'GASPCLDQMFRW'\n",
    "S13 = 'AASPCLDQMFRY'\n",
    "S14 = 'SASPCLDQMFRY'\n",
    "S15 = 'PASPCLDQMFRY'\n",
    "S16 = 'GASPCADQMFRY'\n",
    "S17 = 'GASPCSDQMFRY'\n",
    "S18 = 'GASPCPDQMFRY'\n",
    "S19 = 'GASPCLDQMFAY'\n",
    "S20 = 'GASPCLDQMFSY'\n",
    "S21 = 'GASPCLDQMFPY'\n",
    "S22 = '-ASPCLDQMFRY'\n",
    "S23 = 'G-SPCLDQMFRY'\n",
    "S24 = 'GA-PCLDQMFRY'\n",
    "S25 = 'GAS-CLDQMFRY'\n",
    "S26 = 'GASP-LDQMFRY'\n",
    "S27 = 'GASPC-DQMFRY'\n",
    "S28 = 'GASPCL-QMFRY'\n",
    "S29 = 'GASPCLD-MFRY'\n",
    "S30 = 'GASPCLDQ-FRY'\n",
    "S31 = 'GASPCLDQM-RY'\n",
    "S32 = 'GASPCLDQMF-Y'\n",
    "S33 = 'GASPCLDQMFR-'\n",
    "S34 = 'GASPCLDQMF--'\n",
    "S35 = 'GASPCLDQM---'\n",
    "S36 = 'GASPCLDQ----'\n",
    "S37 = 'GASPCLD-----'\n",
    "S38 = 'GASPCL------'\n",
    "S39 = 'GASPC-------'\n",
    "S40 = 'GASP--------'\n",
    "S41 = '--------MFRY'\n",
    "S42 = '-------QMFRY'\n",
    "S43 = '------DQMFRY'\n",
    "S44 = '-----LDQMFRY'\n",
    "S45 = '----CLDQMFRY'\n",
    "S46 = '---PCLDQMFRY'\n",
    "S47 = '--SPCLDQMFRY'\n",
    "S48 = 'YAS-CLWSMF-Y'\n",
    "S49 = 'G-WP-LDQGFRW'\n",
    "S50 = '-ASPWLDAMAR-'\n",
    "S51 = 'G--PCLDPAFRS'\n",
    "S52 = 'WASPCYDQMF--'\n",
    "S53 = 'GYSPCDL--FRY'\n",
    "S54 = 'GYSPCLD--FRY'\n",
    "S55 = 'GRSPCLD--FRY'\n",
    "S56 = 'GFSPCLD--FRY'\n",
    "S57 = '--WPCLDQPPRY'\n",
    "S58 = '--FPCLDQMYRF'\n",
    "S59 = '--RPCLDQMRFY'\n",
    "S60 = '--YPCLDQMRYF'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34df900e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3dc0013d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   SEQ ID           SEQ        KC      PSIM     PDSIM    KCALT1    KCALT2\n",
      "0      S1  WASPCLDQMFRY  0.965142  0.924694  0.075306  0.965142  1.037470\n",
      "1      S2  GWSPCLDQMFRY  0.971213  0.937778  0.062222  0.971213  1.030545\n",
      "2      S3  GAWPCLDQMFRY  0.977221  0.950739  0.049261  0.977221  1.023866\n",
      "3      S4  GASWCLDQMFRY  0.981433  0.959835  0.040165  0.981433  1.019283\n",
      "4      S5  GASPWLDQMFRY  0.984532  0.966529  0.033471  0.984532  1.015962\n",
      "5      S6  GASPCWDQMFRY  0.988017  0.974063  0.025937  0.988017  1.012277\n",
      "6      S7  GASPCLWQMFRY  0.989984  0.978317  0.021683  0.989984  1.010220\n",
      "7      S8  GASPCLDWMFRY  0.993133  0.985129  0.014871  0.993133  1.006963\n",
      "8      S9  GASPCLDQWFRY  0.994777  0.988688  0.011312  0.994777  1.005278\n",
      "9     S10  GASPCLDQMWRY  0.997201  0.993938  0.006062  0.997201  1.002814\n",
      "10    S11  GASPCLDQMFWY  0.998559  0.996878  0.003122  0.998559  1.001445\n",
      "11    S12  GASPCLDQMFRW  0.999446  0.998799  0.001201  0.999446  1.000555\n",
      "12    S13  AASPCLDQMFRY  0.995955  0.991239  0.008761  0.995955  1.004078\n",
      "13    S14  SASPCLDQMFRY  0.991415  0.981413  0.018587  0.991415  1.008735\n",
      "14    S15  PASPCLDQMFRY  0.988621  0.975369  0.024631  0.988621  1.011644\n",
      "15    S16  GASPCADQMFRY  1.007287  0.984447  0.015553  0.992817  1.007287\n",
      "16    S17  GASPCSDQMFRY  1.004476  0.990392  0.009608  0.995564  1.004476\n",
      "17    S18  GASPCPDQMFRY  1.002741  0.994095  0.005905  0.997274  1.002741\n",
      "18    S19  GASPCLDQMFAY  1.004260  0.990852  0.009148  0.995776  1.004260\n",
      "19    S20  GASPCLDQMFSY  1.003437  0.992606  0.007394  0.996586  1.003437\n",
      "20    S21  GASPCLDQMFPY  1.002928  0.993694  0.006306  0.997089  1.002928\n",
      "21    S22  -ASPCLDQMFRY  1.022896  0.952652  0.047348  0.978107  1.022896\n",
      "22    S23  G-SPCLDQMFRY  1.025072  0.948375  0.051625  0.976125  1.025072\n",
      "23    S24  GA-PCLDQMFRY  1.027083  0.944453  0.055547  0.974308  1.027083\n",
      "24    S25  GAS-CLDQMFRY  1.026776  0.945051  0.054949  0.974585  1.026776\n",
      "25    S26  GASP-LDQMFRY  1.025050  0.948417  0.051583  0.976145  1.025050\n",
      "26    S27  GASPC-DQMFRY  1.023779  0.950911  0.049089  0.977301  1.023779\n",
      "27    S28  GASPCL-QMFRY  1.020643  0.957118  0.042882  0.980175  1.020643\n",
      "28    S29  GASPCLD-MFRY  1.018933  0.960534  0.039466  0.981757  1.018933\n",
      "29    S30  GASPCLDQ-FRY  1.015420  0.967630  0.032370  0.985041  1.015420\n",
      "30    S31  GASPCLDQM-RY  1.012842  0.972900  0.027100  0.987479  1.012842\n",
      "31    S32  GASPCLDQMF-Y  1.009022  0.980814  0.019186  0.991138  1.009022\n",
      "32    S33  GASPCLDQMFR-  1.004683  0.989951  0.010049  0.995360  1.004683\n",
      "33    S34  GASPCLDQMF--  1.014678  0.969141  0.030859  0.985740  1.014678\n",
      "34    S35  GASPCLDQM---  1.030659  0.937561  0.062439  0.971113  1.030659\n",
      "35    S36  GASPCLDQ----  1.052483  0.897494  0.102506  0.952502  1.052483\n",
      "36    S37  GASPCLD-----  1.083218  0.846335  0.153665  0.928656  1.083218\n",
      "37    S38  GASPCL------  1.122433  0.788640  0.211360  0.901650  1.122433\n",
      "38    S39  GASPC-------  1.176035  0.720935  0.279065  0.869803  1.176035\n",
      "39    S40  GASP--------  1.244903  0.648644  0.351356  0.835614  1.244903\n",
      "40    S41  --------MFRY  1.336665  0.571267  0.428733  0.798805  1.336665\n",
      "41    S42  -------QMFRY  1.270864  0.624834  0.375166  0.824311  1.270864\n",
      "42    S43  ------DQMFRY  1.216093  0.677145  0.322855  0.849116  1.216093\n",
      "43    S44  -----LDQMFRY  1.166133  0.732596  0.267404  0.875300  1.166133\n",
      "44    S45  ----CLDQMFRY  1.122999  0.787863  0.212137  0.901285  1.122999\n",
      "45    S46  ---PCLDQMFRY  1.084268  0.844685  0.155315  0.927886  1.084268\n",
      "46    S47  --SPCLDQMFRY  1.050547  0.900915  0.099085  0.954093  1.050547\n",
      "47    S48  YAS-CLWSMF-Y  0.997201  0.993938  0.006062  0.997201  1.002814\n",
      "48    S49  G-WP-LDQGFRW  1.032498  0.934052  0.065948  0.969485  1.032498\n",
      "49    S50  -ASPWLDAMAR-  1.025637  0.947270  0.052730  0.975614  1.025637\n",
      "50    S51  G--PCLDPAFRS  1.073549  0.861817  0.138183  0.935883  1.073549\n",
      "51    S52  WASPCYDQMF--  0.966000  0.926542  0.073458  0.966000  1.036480\n",
      "52    S53  GYSPCDL--FRY  1.008160  0.982615  0.017385  0.991971  1.008160\n",
      "53    S54  GYSPCLD--FRY  1.008213  0.982504  0.017496  0.991920  1.008213\n",
      "54    S55  GRSPCLD--FRY  1.010272  0.978209  0.021791  0.989934  1.010272\n",
      "55    S56  GFSPCLD--FRY  1.012947  0.972685  0.027315  0.987380  1.012947\n",
      "56    S57  --WPCLDQPPRY  1.030863  0.937171  0.062829  0.970932  1.030863\n",
      "57    S58  --FPCLDQMYRF  1.032318  0.934394  0.065606  0.969644  1.032318\n",
      "58    S59  --RPCLDQMRFY  1.030437  0.937986  0.062014  0.971310  1.030437\n",
      "59    S60  --YPCLDQMRYF  1.028090  0.942503  0.057497  0.973404  1.028090\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Analysis (optinalysis) of the well-structured protein sequences\n",
    "protein_sequences = [S1, S2, S3, S4, S5, S6, S7, S8, S9, S10, S11, S12, S13, S14, S15, S16, S17, S18, S19, S20, S21, S22, S23, S24, S25, S26, S27, S28, S29, S30, S31, S32, S33, S34, S35, S36, S37, S38, S39, S40, S41, S42, S43, S44, S45, S46, S47, S48, S49, S50, S51, S52, S53, S54, S55, S56, S57, S58, S59, S60]\n",
    "kc_results = [pairwise_analysis([S0, seq, \"seq_type:protein\", \"pairing:H_H\", \"print:kc\"]) for seq in protein_sequences]\n",
    "psim_results = [pairwise_analysis([S0, seq, \"seq_type:protein\", \"pairing:H_H\", \"print:psim\"]) for seq in protein_sequences]\n",
    "pdsim_results = [pairwise_analysis([S0, seq, \"seq_type:protein\", \"pairing:H_H\", \"print:pdsim\"]) for seq in protein_sequences]\n",
    "kcalt1_results = [pairwise_analysis([S0, seq, \"seq_type:protein\", \"pairing:H_H\", \"print:kcalt1\"]) for seq in protein_sequences]\n",
    "kcalt2_results = [pairwise_analysis([S0, seq, \"seq_type:protein\", \"pairing:H_H\", \"print:kcalt2\"]) for seq in protein_sequences]\n",
    "\n",
    "seq_ID = ['S{}'.format(i) for i in range(1, len(protein_sequences) + 1)]\n",
    "protein_seq = [str(i) for i in protein_sequences]\n",
    "df = pd.DataFrame({\"SEQ ID\":seq_ID, \"SEQ\":protein_seq, \"KC\":kc_results, \"PSIM\":psim_results, \"PDSIM\":pdsim_results, \"KCALT1\":kcalt1_results, \"KCALT2\":kcalt2_results})\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db49abe2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
